datafusion/src/scalar.rs

```diff
@@ -277,6 +277,31 @@ impl std::hash::Hash for ScalarValue {
     }
 }
 
+// return the index into the dictionary values for array@index as well
+// as a reference to the dictionary values array. Returns None for the
+// index if the array is NULL at index
```

Comment:
https://github.com/apache/arrow-rs/issues/672 proposes adding this upstream in arrow

I think this properly handles `null` values now in the DictionaryArray, whereas y initial version did not